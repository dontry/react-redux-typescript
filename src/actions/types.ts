import { Customer } from "../models";

export const SELECT_CUSTOMER = "SELECT_CUSTOMER";
export const ADD_CUSTOMER = "ADD_CUSTOMER";
export const UPDATE_CUSTOMER = "UPDATE_CUSTOMER";
export const DELETE_CUSTOMER = "DELETE_CUSTOMER";
export const TOGGLE_EDITOR_DIALOG = "TOGGLE_EDITOR_DIALOG";
export const UPDATE_QUERY = "UPDATE_QUERY";

export interface SelectCustomerAction {
  type: typeof SELECT_CUSTOMER;
  payload: string | null;
}

export interface AddCustomerAction {
  type: typeof ADD_CUSTOMER;
  payload: Customer;
}

export interface DeleteCustomerAction {
  type: typeof DELETE_CUSTOMER;
  payload: string;
}

export interface UpdateCustomerAction {
  type: typeof UPDATE_CUSTOMER;
  payload: {
    id: string;
    customer: Customer;
  };
}

export interface ToggleEditorDialogAtion {
  type: typeof TOGGLE_EDITOR_DIALOG;
  payload: boolean;
}

export interface UpdateQueryAction {
  type: typeof UPDATE_QUERY;
  payload: string;
}
