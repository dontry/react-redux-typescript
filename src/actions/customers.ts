import {
  ADD_CUSTOMER,
  DELETE_CUSTOMER,
  UPDATE_CUSTOMER,
  AddCustomerAction,
  DeleteCustomerAction,
  UpdateCustomerAction,
  UpdateQueryAction,
  SelectCustomerAction,
  SELECT_CUSTOMER,
  UPDATE_QUERY
} from "./types";
import { Customer } from "../models";

export function selectCustomer(id: string | null): SelectCustomerAction {
  return {
    type: SELECT_CUSTOMER,
    payload: id
  };
}

export function addCustomer(newCustomer: Customer): AddCustomerAction {
  return {
    type: ADD_CUSTOMER,
    payload: newCustomer
  };
}

export function updateCustomer(id: string, customer: Customer): UpdateCustomerAction {
  return {
    type: UPDATE_CUSTOMER,
    payload: { id, customer }
  };
}

export function deleteCustomer(id: string): DeleteCustomerAction {
  return {
    type: DELETE_CUSTOMER,
    payload: id
  };
}

export function updateQuery(query: string): UpdateQueryAction {
  return {
    type: UPDATE_QUERY,
    payload: query
  };
}
