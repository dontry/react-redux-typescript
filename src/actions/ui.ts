import { ToggleEditorDialogAtion, TOGGLE_EDITOR_DIALOG } from "./types";

export function toggleEditorDialog(isOpen: boolean): ToggleEditorDialogAtion {
  return {
    type: TOGGLE_EDITOR_DIALOG,
    payload: isOpen
  };
}
