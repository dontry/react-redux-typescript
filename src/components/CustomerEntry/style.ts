import styled from "styled-components";

export const EntryWrapper = styled.li`
  display: flex;
  justify-content: space-between;
  padding: 8px 16px;
`;

export const StyledName = styled.span`
  font-size: 1em;
`;
