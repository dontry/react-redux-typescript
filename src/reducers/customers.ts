import { Customer } from "../models";
import { createSelector } from "reselect";
import {
  ADD_CUSTOMER,
  DELETE_CUSTOMER,
  UPDATE_CUSTOMER,
  SELECT_CUSTOMER,
  UPDATE_QUERY
} from "../actions/types";
import { CustomerState, RootState } from "./state";
import { combineReducers } from "redux";

export const initialState: CustomerState = {
  selectedId: null,
  query: "",
  list: []
};

export const selectedId = (state = initialState.selectedId, action: any) => {
  switch (action.type) {
    case SELECT_CUSTOMER:
      return action.payload;
    default:
      return state;
  }
};

export const query = (state = initialState.query, action: any) => {
  switch (action.type) {
    case UPDATE_QUERY:
      return action.payload;
    default:
      return state;
  }
};

export const list = (state = initialState.list, action: any) => {
  switch (action.type) {
    case ADD_CUSTOMER:
      return [...state, action.payload];
    case DELETE_CUSTOMER:
      return state.filter(customer => customer.id !== action.payload);
    case UPDATE_CUSTOMER:
      return state.map(customer =>
        customer.id === action.payload.id ? action.payload.customer : customer
      );
    default:
      return state;
  }
};

export const customerReducer = combineReducers({
  selectedId,
  list,
  query
});

// Selector
export const selectCustomers = (state: any): Customer[] => state.customers.list;
export const selectQuery = (state: any): string => state.customers.query;
export const selectCustomerById = createSelector(
  selectCustomers,
  state => state.customers.selectedId,
  (customers: Customer[], id: string | null) => customers.find(customer => customer.id === id)
);

export const selectCustomerByQuery = createSelector(
  selectCustomers,
  selectQuery,
  (customers: Customer[], query: string) => {
    if (query === "") {
      return customers;
    } else {
      const regex: RegExp = new RegExp(query, "i");
      return customers.filter(customer => {
        const fullname = `${customer.firstName} ${customer.lastName}`;
        return regex.test(fullname);
      });
    }
  }
);
