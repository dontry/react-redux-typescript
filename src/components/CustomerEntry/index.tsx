import React from "react";
import { Customer } from "../../models";
import { EntryWrapper, ButtonGroup } from "./style";
import { deleteCustomer } from "../../actions/customers";
import { Button } from "@material-ui/core";

interface Props {
  customer: Customer;
  onDelete?: typeof deleteCustomer;
}

const CustomerEntry = ({ customer, onDelete }: Props) => {
  const fullName = `${customer.firstName} ${customer.lastName}`;

  const _handleDelete = () => {
    onDelete!(customer.id);
  };

  return (
    <EntryWrapper>
      <span data-testid="customer-name">{fullName}</span>
      <ButtonGroup>
        <Button color="primary">Edit</Button>
        {onDelete && (
          <Button data-testid="delete-btn" color="secondary" onClick={_handleDelete}>
            Delete
          </Button>
        )}
      </ButtonGroup>
    </EntryWrapper>
  );
};

export default CustomerEntry;
