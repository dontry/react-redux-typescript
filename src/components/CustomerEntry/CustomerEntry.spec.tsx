import React from "react";
import CustomerEntry from "./index";
import { render, cleanup, fireEvent } from "../../test/fixtures/config";
import { createCustomer } from "../../test/fixtures/customerMock";
import * as actions from "../../actions/customers";

afterEach(cleanup);

describe("CustomerEntry", () => {
  it("should render the component with a customer name", () => {
    const customer = createCustomer();
    const { getByTestId } = render(<CustomerEntry customer={customer} />);
    expect(getByTestId("customer-name")).toHaveTextContent(
      `${customer.firstName} ${customer.lastName}`
    );
  });

  it("calls deleteCustomer function by clicking delete button", () => {
    const spy = jest.spyOn(actions, "deleteCustomer");
    const customer = createCustomer();

    const { getByTestId } = render(
      <CustomerEntry customer={customer} onDelete={actions.deleteCustomer} />
    );

    fireEvent.click(getByTestId("delete-btn"));

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(customer.id);
  });
});
