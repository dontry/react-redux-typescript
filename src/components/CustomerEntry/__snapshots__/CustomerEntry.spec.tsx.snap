// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerEntry should render a customer 1`] = `
<li
  class="sc-bdVaJa bBrGBh"
>
  <span
    data-testid="customer-name"
  >
    Tatyana Johns
  </span>
  <span
    class="sc-bwzfXH hLvLiK"
  />
</li>
`;
